lib/msf/core/exploit/remote/http/exchange.rb

```diff
@@ -0,0 +1,122 @@
+          #
+          # @return [Rex::Version, nil] The Exchange version if it was able to be recovered. Nil otherwise
+          def exchange_get_version
+            # First lets try a cheap way of doing this via a leak of the X-OWA-Version header.
+            # If we get this we know the version number for sure and we can skip a lot of leg work.
+            res = send_request_cgi(
+              'method' => 'GET',
+              'uri' => normalize_uri(target_uri.path, '/owa/service')
+            )
+
+            unless res
+              print_error('Target did not respond!')
+              return nil
+            end
+
+            if res.headers['X-OWA-Version']
+              build = res.headers['X-OWA-Version']
+              return Rex::Version.new(build)
+            end
+
+            # Next, determine if we are up against an older version of Exchange Server where
+            # the /owa/auth/logon.aspx page gives the full version. Recent versions of Exchange
+            # give only a partial version without the build number.
+            res = send_request_cgi(
+              'method' => 'GET',
+              'uri' => normalize_uri(target_uri.path, '/owa/auth/logon.aspx')
+            )
+
+            unless res
+              print_error('Target did not respond!')
+              return nil
+            end
+
+            if res.code == 200 && ((%r{/owa/(?<build>\d+\.\d+\.\d+\.\d+)} =~ res.body) || (%r{/owa/auth/(?<build>\d+\.\d+\.\d+\.\d+)} =~ res.body))
+              return Rex::Version.new(build)
+            end
+
+            # Next try @tseller's way and try /ecp/Current/exporttool/microsoft.exchange.ediscovery.exporttool.application
+            # URL which if successful should provide some XML with entries like the following:
+            #
+            # <assemblyIdentity name="microsoft.exchange.ediscovery.exporttool.application"
+            # version="15.2.986.5" publicKeyToken="b1d1a6c45aa418ce" language="neutral"
+            # processorArchitecture="msil" xmlns="urn:schemas-microsoft-com:asm.v1" />
+            #
+            # This only works on Exchange Server 2013 and later and may not always work, but if it
+            # does work it provides the full version number so its a nice strategy.
+            res = send_request_cgi(
+              'method' => 'GET',
+              'uri' => normalize_uri(target_uri.path, '/ecp/current/exporttool/microsoft.exchange.ediscovery.exporttool.application')
+            )
+
+            unless res
+              print_error('Target did not respond!')
+              return nil
+            end
+
+            if res.code == 200 && res.body =~ /name="microsoft.exchange.ediscovery.exporttool" version="\d+\.\d+\.\d+\.\d+"/
+              build = res.body.match(/name="microsoft.exchange.ediscovery.exporttool" version="(\d+\.\d+\.\d+\.\d+)"/)[1]
+              return Rex::Version.new(build)
```

Comment:
You can do the same named capture here as you do in the check above on L62 to skip applying the same regex a second time.

```suggestion
            if res.code == 200 && res.body =~ /name="microsoft.exchange.ediscovery.exporttool" version="(?<build>\d+(\.\d+){3})"/
              return Rex::Version.new(Regexp.last_match('build'))            
```

Reply:
Suggested change made in https://github.com/rapid7/metasploit-framework/pull/16249/commits/08032259599322475fb8d7db28c0c4e043c49fdc